### ISSUE TITLE
Parsing empty array to string results in null and not an exception

### ISSUE THREAD

**Issue by BrennanConroy** [dotnet/corefx#36901]
```c#
var json = "[]";
var output = JsonSerializer.Parse<string>(json); // should throw
// output == null here
```

cc @steveharter

**Comment by steveharter** [comment]
Fwiw the code in question does check if the type (in this case `typeof(string)`) implements `IEnumerable` and since it does the code doesn't throw. So another check needs to be made...

**Comment by BrennanConroy** [comment]
Looks like this hits a Debug.Assert at https://github.com/dotnet/corefx/blob/master/src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleArray.cs#L41 which fails some of our tests when run locally.

**Comment by ahsonkhan** [comment]
Marking as up-for-grabs if anyone would like to pick this up.

**Comment by watfordgnf** [comment]
This looks fixed on current master 189766719a059512a515d488efc5c65c7abf0650. I can put up a PR with additional tests to ensure this dose not regress.

**Comment by ahsonkhan** [comment]
> Fwiw the code in question does check if the type (in this case `typeof(string)`) implements `IEnumerable` and since it does the code doesn't throw. So another check needs to be made...

> This looks fixed on current master 189766719a059512a515d488efc5c65c7abf0650.

@steveharter, where was this fixed? Can you link to the relevant code you were referring to that needed additional type checks (is it here): https://github.com/dotnet/corefx/blob/3ff5c4a6eb6315c61cfb0078245c31b59f798290/src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleArray.cs#L45-L79)

> Looks like this hits a Debug.Assert at https://github.com/dotnet/corefx/blob/master/src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleArray.cs#L41 which fails some of our tests when run locally.

@BrennanConroy, imo, it's helpful if you push a permalink so that if the code were to change, we can still refer back to the original point-in-time commit/code. For this case, looking at the date of your post, I believe you were pointing to this Debug.Assert:
https://github.com/dotnet/corefx/blob/66bcc2d2b1248ed5384393adf5f3a647d18b75f1/src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleArray.cs#L41

Looks like this issue has been fixed. Thanks for the new tests @watfordgnf.
